File: CS39000-CP2/competition2/Keyboard.java
import java.util.*;
import java.io.*;
import java.lang.*;

public class Keyboard {
  static class  Node {
    public char data;
    public Node next;

    Node() {
      data = '\0';
      next = null;
    }

    Node(char c, Node n) {
      data = c;
      next = n;
    }
  }

  public static void main(String[] args) throws IOException {
    Scanner scan = new Scanner(System.in);
    // FastReader fr = new FastReader(System.in);
    PrintWriter out = new PrintWriter(System.out);
		
    while (true) {
      String s = scan.nextLine();
      if (s.equals("0")) {
        break; 
      }

      Node curr = new Node();
      Node start = curr;
      Node end = curr;

      for (int i = 0; i < s.length(); i++) {
        if (s.charAt(i) == '@') {
          curr = start;
        } else if (s.charAt(i) == '$') {
          curr = end;
        } else {
          Node temp = new Node(s.charAt(i), curr.next);
          curr.next = temp;

          end = curr == end ? temp : end;
          curr = temp;
        }
      }

      while (start != null) {
        if (start.data != '\0') {
          out.print(start.data);
        }
        start = start.next;
      }
      out.println();
    }
    out.close();
  }
}

File: CS39000-CP2/competition2/keyboard.cpp
#include <bits/stdc++.h>
#include <iostream>
#include <stdio.h>
#include <tuple>
#include <utility>

using namespace std;
string s;

struct node {
  char data;
  struct node *next;
};

node newNode(char c, node *n) {
  node *n = new node;
  n->data = c;
  n->next = n;

  return n;
}

int main() {
  while (true) {
    cin >> s;
    if (s == "0") { break; }

    node *curr = new node;
    curr->data = '\0';
    curr->next = NULL;
    node *start = curr;
    node *end = curr;

    for (int i = 0; i < s.size(); i++) {
      if (s[i] == '@') {
        curr = start;
        continue;
      }
      if (s[i] == '$') {
        curr = end;
        continue;
      }

      node *temp = new node;
      temp->data = s[i];
      
      temp->next = curr->next;
      curr->next = temp;

      end = curr == end ? temp : end;
      curr = temp;
    }

    while (start != NULL) {
      if (start->data != '\0') {
        cout << start->data;
      }
      start = start->next;
    }
    cout << endl;
  }

  return 0;
}